From 5fd0407441dcd5810a6a6b560564ddaf68eff9f1 Mon Sep 17 00:00:00 2001
From: Gu Chao <gux.chao@intel.com>
Date: Sun, 12 Apr 2020 19:33:02 +0800
Subject: [PATCH] DRV_DPM_SW-182: DPM Update EPON continuous Q allocation with
 new PPv4 API

- code refactoring to separate queue and scheduler handling
- add flag_ops to struct resv_info
---
 drivers/net/datapath/dpm/gswip31/datapath_misc.h   |   5 +-
 .../net/datapath/dpm/gswip31/datapath_ppv4_api.c   | 341 ++++++++++-----------
 2 files changed, 161 insertions(+), 185 deletions(-)

diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.h b/drivers/net/datapath/dpm/gswip31/datapath_misc.h
index 608798a950c9..b4a361cf1cac 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.h
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.h
@@ -83,7 +83,8 @@ struct resv_sch {
 struct resv_info {
 	int num_resv_q; /*!< input:reserve the required number of queues*/
 	int num_resv_sched; /*!< input:reserve required number of schedulers*/
-	struct resv_q  *resv_q; /*!< reserved queues info*/
+	int flag_ops; /*!< saved op flags of enum DP_DEV_DATA_FLAG */
+	struct resv_q *resv_q; /*!< reserved queues info*/
 	struct resv_sch *resv_sched; /*!< reserved schedulers info */
 };
 
@@ -246,8 +247,6 @@ int dp_get_gsw_pmapper_31(int inst, int bport, int lport,
 int dp_children_get_31(struct dp_node_child *cfg, int flag);
 int dp_free_children_via_parent_31(struct dp_node_alloc *node, int flag);
 int dp_node_reserve(int inst, int ep, struct dp_dev_data *data, int flags);
-int dp_q_reserve_continuos(int inst, int ep, struct dp_dev_data *data,
-			   int flags);
 int dp_qos_level_get_31(struct dp_qos_level *dp, int flag);
 int dp_meter_alloc_31(int inst, int *meterid, int flag);
 int dp_meter_add_31(struct net_device *dev,
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
index a0a36e79c5c4..15b7d2ecb029 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
@@ -4158,218 +4158,195 @@ int dp_children_get_31(struct dp_node_child *cfg, int flag)
 	return DP_FAILURE;
 }
 
-static int dp_queue_alloc_conf(int inst, int *logical_id, int *qid)
+static int dp_q_reserve_contiguous(int inst, int ep, struct dp_dev_data *data)
 {
-	u32 id = 0;
 	struct hal_priv *priv = HAL(inst);
-	struct pp_qos_queue_info info;
-	int res = DP_SUCCESS;
+	struct resv_info *r = &priv->resv[ep];
+	size_t len;
+	int i, res;
+	u32 *ids, *phy_ids;
+
+	len = sizeof(*ids) * data->num_resv_q * 2;
+	ids = devm_kzalloc(&g_dp_dev->dev, len, GFP_ATOMIC);
+	if (!ids)
+		return DP_FAILURE;
 
-	if (dp_qos_queue_allocate_id_phy(priv->qdev, &id, &info.physical_id)) {
-		pr_err("qos_queue_allocate failed\n");
+	phy_ids = ids + data->num_resv_q;
+	res = pp_qos_contiguous_queue_allocate(priv->qdev, ids, phy_ids,
+					       data->num_resv_q);
+	if (res) {
+		pr_err("%s: pp qos contiguous q alloc failed with %d\n",
+		       __func__, res);
+		devm_kfree(&g_dp_dev->dev, ids);
 		return DP_FAILURE;
 	}
-
-	*qid = info.physical_id;
-	*logical_id = id;
-	return res;
+	for (i = 0; i < data->num_resv_q; i++) {
+		r->resv_q[i].id = ids[i];
+		r->resv_q[i].physical_id = phy_ids[i];
+		DP_DEBUG(DP_DBG_FLAG_QOS, "%s: reseve q[%d/%d] for ep %d\n",
+			 __func__, ids[i], phy_ids[i], ep);
+	}
+	r->num_resv_q = data->num_resv_q;
+	r->flag_ops |= DP_F_DEV_CONTINUOUS_Q;
+	data->qos_resv_q_base =	phy_ids[0];
+	devm_kfree(&g_dp_dev->dev, ids);
+	return DP_SUCCESS;
 }
 
-int dp_q_reserve_continuous(int inst, int ep, struct dp_dev_data *data,
-			    int flags)
+static int dp_q_remove(int inst, int ep)
 {
-	int i;
-	u32 id = 0, qid = 0;
-	int len, curr_num = 0, curr_off = 0;
-	struct resv_q *resv_q = NULL, *tmp_resv_q = NULL;
 	struct hal_priv *priv = HAL(inst);
-	int res = DP_SUCCESS;
+	struct resv_info *r = &priv->resv[ep];
+	int i, res;
 
-	/* free resved queue */
-	if (flags == DP_F_DEREGISTER)
-		goto FREE_EXIT;
-	/* to reserve the queue */
-	if (data->num_resv_q <= 0) {
-		pr_err("Provide valid Q no for continuous Q reserve\n");
-		return DP_FAILURE;
-	}
+	if (!r->resv_q)
+		return DP_SUCCESS;
 
-	len = sizeof(struct resv_q) * data->num_resv_q;
-	priv->resv[ep].resv_q = devm_kzalloc(&g_dp_dev->dev, len, GFP_ATOMIC);
-	if (!priv->resv[ep].resv_q)
-		return DP_FAILURE;
-	tmp_resv_q = devm_kzalloc(&g_dp_dev->dev,
-				  sizeof(*tmp_resv_q) * MAX_QUEUE, GFP_ATOMIC);
-	if (!tmp_resv_q) {
-		devm_kfree(&g_dp_dev->dev, priv->resv[ep].resv_q);
-		return DP_FAILURE;
-	}
-	if (dp_queue_alloc_conf(inst, &id, &qid)) {
-		pr_err("qos_queue_allocate failed\n");
-		devm_kfree(&g_dp_dev->dev, priv->resv[ep].resv_q);
-		devm_kfree(&g_dp_dev->dev, tmp_resv_q);
-		return DP_FAILURE;
-	}
-	tmp_resv_q[curr_num].id = id;
-	tmp_resv_q[curr_num].physical_id = qid;
-	curr_num++;
-	while (((curr_num) - curr_off) < data->num_resv_q) {
-		if (dp_queue_alloc_conf(inst, &id, &qid)) {
-			res = DP_FAILURE;
-			pr_err("qos_queue_allocate failed\n");
-			devm_kfree(&g_dp_dev->dev, priv->resv[ep].resv_q);
-			goto clear_q;
-		}
-		tmp_resv_q[curr_num].id = id;
-		tmp_resv_q[curr_num].physical_id = qid;
-		curr_num++;
-		if ((tmp_resv_q[curr_num - 1].physical_id -
-		     tmp_resv_q[curr_num - 2].physical_id == 1) ||
-		    (tmp_resv_q[curr_num - 2].physical_id -
-		     tmp_resv_q[curr_num - 1].physical_id == 1)) {
-			DP_DEBUG(DP_DBG_FLAG_QOS, "%d %d\n",
-				 curr_num, curr_off);
-			continue;
-		} else {
-			curr_off = curr_num - 1;
-		}
-	}
-	if (curr_num - curr_off == data->num_resv_q) {
-		/* Copy information to priv structure */
-		resv_q = priv->resv[ep].resv_q;
-		data->qos_resv_q_base =	tmp_resv_q[curr_off].physical_id;
-		for (i = 0; i < data->num_resv_q; i++) {
-			resv_q[i].id = tmp_resv_q[curr_off + i].id;
-			resv_q[i].physical_id =
-				tmp_resv_q[curr_off + i].physical_id;
-			priv->resv[ep].num_resv_q = i + 1;
-			DP_DEBUG(DP_DBG_FLAG_QOS, "reseve q[%d/%d]\n",
-				 resv_q[i].id, resv_q[i].physical_id);
-		}
-	}
-clear_q:
-	/* clear other non-continuously allocated Q */
-	for (i = 0; i < curr_off; i++)
-		dp_qos_queue_remove(priv->qdev, tmp_resv_q[i].id);
-	devm_kfree(&g_dp_dev->dev, tmp_resv_q);
-	return res;
-FREE_EXIT:
-	if (priv->resv[ep].resv_q) {
-		struct resv_q  *resv_q = priv->resv[ep].resv_q;
+	if (r->num_resv_q <= 0)
+		goto out;
 
-		for (i = 0; i < priv->resv[ep].num_resv_q; i++)
-			dp_qos_queue_remove(priv->qdev, resv_q[i].id);
-		devm_kfree(&g_dp_dev->dev, resv_q);
-		priv->resv[ep].resv_q = NULL;
-		priv->resv[ep].num_resv_q = 0;
+	if (r->flag_ops & DP_F_DEV_CONTINUOUS_Q) {
+		res = pp_qos_contiguous_queue_remove(priv->qdev,
+						     r->resv_q[0].id,
+						     r->num_resv_q);
+		if (res) {
+			pr_err("%s: pp qos contiguous q remove failed with %d\n",
+			       __func__, res);
+			return DP_FAILURE;
+		}
+		r->num_resv_q = 0;
+		r->flag_ops &= ~DP_F_DEV_CONTINUOUS_Q;
+		goto out;
 	}
-	return res;
+	while (r->num_resv_q > 0) {
+		i = r->num_resv_q - 1;
+		res = dp_qos_queue_remove(priv->qdev, r->resv_q[i].id);
+		if (res) {
+			pr_err("%s: pp qos q remove failed with %d for q id %u\n",
+			       __func__, res, r->resv_q[i].id);
+			return DP_FAILURE;
+		}
+		r->num_resv_q--;
+	}
+out:
+	devm_kfree(&g_dp_dev->dev, r->resv_q);
+	r->resv_q = NULL;
+	return DP_SUCCESS;
 }
 
-int dp_node_reserve(int inst, int ep, struct dp_dev_data *data, int flags)
+static int dp_q_reserve(int inst, int ep, struct dp_dev_data *data)
 {
-	int i;
-	u32 id;
-	struct pp_qos_queue_info info;
-	int len;
-	struct resv_q *resv_q = NULL;
-#ifdef WORKAROUND_DROP_PORT
-	struct pp_qos_queue_conf q_conf;
-#endif
 	struct hal_priv *priv = HAL(inst);
-	int res = DP_SUCCESS;
-
-	if (!priv) {
-		pr_err("priv cannot be NULL\n");
-		return DP_FAILURE;
-	}
+	struct resv_info *r = &priv->resv[ep];
+	size_t len;
+	int i, res;
+	u32 id, phy_id;
 
-	/* free resved queue/scheduler */
-	if (flags == DP_F_DEREGISTER)
-		goto FREE_EXIT;
-
-	if (data->flag_ops & (DP_F_DEV_RESV_Q | DP_F_DEV_CONTINUOUS_Q)) {
-		dp_q_reserve_continuous(inst, ep, data, flags);
-		return res;
-	}
+	if (!(data->flag_ops & DP_F_DEV_RESV_Q) || data->num_resv_q <= 0)
+		return DP_SUCCESS;
 
-	/* to reserve the queue */
-	if (data->num_resv_q <= 0)
-		goto RESERVE_SCHED;
 	len = sizeof(struct resv_q) * data->num_resv_q;
-	priv->resv[ep].resv_q = devm_kzalloc(&g_dp_dev->dev, len, GFP_ATOMIC);
-	if (!priv->resv[ep].resv_q) {
-		res = DP_FAILURE;
-		goto FREE_EXIT;
+	r->resv_q = devm_kzalloc(&g_dp_dev->dev, len, GFP_ATOMIC);
+	if (!r->resv_q)
+		return DP_FAILURE;
+
+	if (data->flag_ops & DP_F_DEV_CONTINUOUS_Q) {
+		if (dp_q_reserve_contiguous(inst, ep, data)) {
+			dp_q_remove(inst, ep);
+			return DP_FAILURE;
+		}
+		return DP_SUCCESS;
 	}
-	DP_DEBUG(DP_DBG_FLAG_QOS, "queue size =%d for ep=%d\n", len, ep);
-	resv_q = priv->resv[ep].resv_q;
 	for (i = 0; i < data->num_resv_q; i++) {
-		if (dp_qos_queue_allocate_id_phy(priv->qdev, &id,
-						 &info.physical_id)) {
-			res = DP_FAILURE;
-			pr_err("qos_queue_allocate failed\n");
-			goto FREE_EXIT;
-		}
-		resv_q[i].id = id;
-		resv_q[i].physical_id = info.physical_id;
-		priv->resv[ep].num_resv_q = i + 1;
-		DP_DEBUG(DP_DBG_FLAG_QOS,
-			 "reseve q[%d/%d]\n",
-			 resv_q[i].id, resv_q[i].physical_id);
-	}
-RESERVE_SCHED:
-	/* reserve scheduler */
-	if (data->num_resv_sched > 0) {
-		int len;
-		struct resv_sch *p_sch = priv->resv[ep].resv_sched;
-
-		len = sizeof(struct resv_sch) * data->num_resv_sched;
-		priv->resv[ep].resv_sched = devm_kzalloc(&g_dp_dev->dev,
-					    len, GFP_ATOMIC);
-		DP_DEBUG(DP_DBG_FLAG_QOS, "sched size =%d for ep=%d\n",
-			 len, ep);
-		if (!priv->resv[ep].resv_sched) {
-			res = DP_FAILURE;
-			goto FREE_EXIT;
+		res = dp_qos_queue_allocate_id_phy(priv->qdev, &id, &phy_id);
+		if (res) {
+			pr_err("%s: pp qos q alloc failed with %d\n",
+			       __func__, res);
+			dp_q_remove(inst, ep);
+			return DP_FAILURE;
 		}
-		p_sch = priv->resv[ep].resv_sched;
-		for (i = 0; i < data->num_resv_sched; i++) {
-			if (dp_qos_sched_allocate(priv->qdev, &id)) {
-				res = DP_FAILURE;
-				pr_err("qos_queue_allocate failed\n");
-				goto FREE_EXIT;
-			}
-			DP_DEBUG(DP_DBG_FLAG_QOS,
-				 "qos_sched_allocate: %d\n", id);
-			p_sch[i].id = id;
-			priv->resv[ep].num_resv_sched = i + 1;
-			DP_DEBUG(DP_DBG_FLAG_QOS,
-				 "reseve sched[/%d]\n", p_sch[i].id);
+		r->resv_q[i].id = id;
+		r->resv_q[i].physical_id = phy_id;
+		r->num_resv_q++;
+		DP_DEBUG(DP_DBG_FLAG_QOS, "%s: reseve q[%d/%d] for ep %d\n",
+			 __func__, id, phy_id, ep);
+	}
+	return DP_SUCCESS;
+}
+
+static int dp_sched_remove(int inst, int ep)
+{
+	struct hal_priv *priv = HAL(inst);
+	struct resv_info *r = &priv->resv[ep];
+	int i, res;
+
+	if (!r->resv_sched)
+		return DP_SUCCESS;
+
+	while (r->num_resv_sched > 0) {
+		i = r->num_resv_sched - 1;
+		res = dp_qos_sched_remove(priv->qdev, r->resv_sched[i].id);
+		if (res) {
+			pr_err("%s: pp qos sched remove failed with %d for sched id %u\n",
+			       __func__, res, r->resv_sched[i].id);
+			return DP_FAILURE;
 		}
+		r->num_resv_sched--;
 	}
-	return res;
+	devm_kfree(&g_dp_dev->dev, r->resv_sched);
+	r->resv_sched = NULL;
+	return DP_SUCCESS;
+}
+
+static int dp_sched_reserve(int inst, int ep, struct dp_dev_data *data)
+{
+	struct hal_priv *priv = HAL(inst);
+	struct resv_info *r = &priv->resv[ep];
+	size_t len;
+	int i, res;
+	u32 id;
+
+	if (!(data->flag_ops & DP_F_DEV_RESV_SCH) || data->num_resv_sched <= 0)
+		return DP_SUCCESS;
 
-FREE_EXIT:
-	if (priv->resv[ep].resv_q) {
-		struct resv_q  *resv_q = priv->resv[ep].resv_q;
+	len = sizeof(struct resv_sch) * data->num_resv_sched;
+	r->resv_sched = devm_kzalloc(&g_dp_dev->dev, len, GFP_ATOMIC);
+	if (!r->resv_sched)
+		return DP_FAILURE;
 
-		for (i = 0; i < priv->resv[ep].num_resv_q; i++)
-			dp_qos_queue_remove(priv->qdev, resv_q[i].id);
-		devm_kfree(&g_dp_dev->dev, resv_q);
-		priv->resv[ep].resv_q = NULL;
-		priv->resv[ep].num_resv_q = 0;
+	for (i = 0; i < data->num_resv_sched; i++) {
+		res = dp_qos_sched_allocate(priv->qdev, &id);
+		if (res) {
+			pr_err("%s: pp qos sched alloc failed with %d\n",
+			       __func__, res);
+			dp_sched_remove(inst, ep);
+			return DP_FAILURE;
+		}
+		r->resv_sched[i].id = id;
+		r->num_resv_sched++;
+		DP_DEBUG(DP_DBG_FLAG_QOS, "%s: reseve sched[%u] for ep %d\n",
+			 __func__, id, ep);
 	}
-	if (priv->resv[ep].resv_sched) {
-		struct resv_sch *resv_sch = priv->resv[ep].resv_sched;
+	return DP_SUCCESS;
+}
 
-		for (i = 0; i < priv->resv[ep].num_resv_sched; i++)
-			dp_qos_sched_remove(priv->qdev, resv_sch[i].id);
-		devm_kfree(&g_dp_dev->dev, resv_sch);
-		priv->resv[ep].resv_sched = NULL;
-		priv->resv[ep].num_resv_sched = 0;
+int dp_node_reserve(int inst, int ep, struct dp_dev_data *data, int flags)
+{
+	int rc0, rc1, rc2;
+
+	if (flags == DP_F_DEREGISTER) {
+		rc0 = DP_SUCCESS;
+		goto remove;
 	}
-	return res;
+	rc0 = dp_q_reserve(inst, ep, data);
+	if (!rc0)
+		rc0 = dp_sched_reserve(inst, ep, data);
+	if (!rc0)
+		return DP_SUCCESS;
+remove:
+	rc1 = dp_sched_remove(inst, ep);
+	rc2 = dp_q_remove(inst, ep);
+	return (rc0 || rc1 || rc2) ? DP_FAILURE : DP_SUCCESS;
 }
 
 /* dp_qos_global_info_get_31 API
